src/Sentry/Protocol/Envelope.cs

```diff
@@ -75,5 +80,45 @@ public static Envelope FromEvent(SentryEvent @event)
 
             return new Envelope(header, items);
         }
+
+        private static async Task<IReadOnlyDictionary<string, object>> DeserializeHeaderAsync(
+            Stream stream,
+            CancellationToken cancellationToken = default)
+        {
+            var buffer = new List<byte>();
+
+            var lastLastByte = default(int);
+            var lastByte = await stream.ReadByteAsync(cancellationToken).ConfigureAwait(false);
+            while (lastByte != -1 && !(lastByte == (byte)'\n' && lastLastByte != (byte)'\\'))
+            {
+                buffer.Add((byte)lastByte);
+
+                lastLastByte = lastByte;
+                lastByte = await stream.ReadByteAsync(cancellationToken).ConfigureAwait(false);
+            }
+
+            return Json.DeserializeFromByteArray<Dictionary<string, object>>(buffer.ToArray());
+        }
+
+        /// <summary>
+        /// Deserializes envelope from stream.
+        /// </summary>
+        public static async Task<Envelope> DeserializeAsync(
+            Stream stream,
+            CancellationToken cancellationToken = default)
+        {
+            // Header
+            var header = await DeserializeHeaderAsync(stream, cancellationToken).ConfigureAwait(false);
```

Comment:
Check if we got something back here? If we got null the thing is broken and we should bail.
Oh I guess this doesn't return null, right? it'll be an empty map.

Reply:
Good point. It does return `null` because Json.NET is nice like that.

Reply:
Should `null` be an exception or coerced to an empty map?

Reply:
Good question. I dunno. `null` kind of sucks, if the header is invalid then maybe we should disconsider the payload instead. Empty map will lead us to keep going and try to read the rest of the stream as envelope item

Reply:
Changed it to exception, looks cleaner now